### PR TITLE
all: implement code completion

### PR DESCRIPTION
**Pull request by l-zeuch**
**Please describe the changes this pull request does and why it should be merged:**
Add completions for use with https://github.com/hrsh7th/nvim-cmp.

Such a feature is incredibly often requested and considering how easy it was to implement this feature, it's only valid to ship it to our users.

Instructions on how to enable code completion are added to the README.
I also added a `make` target to regenerate the sources in-place when needed.
The generated lua code is exempt from language statistics, but I opted to keep `boilerplate.lua` and `yagpdb.lua`.

**Terms**
- [x] I agree to follow this project's [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] I have read and understood this project's [Contributing Guidelines](../CONTRIBUTING.md)

### CODE REVIEW

File: scripts/gen_code_completion.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    _file_str = None
+
+    def __init__(self):
+        self.file_str = StringIO()
+
+    def Add(self, str):
+        self.file_str.write(str)
+
+    def __str__(self):
+        return self.file_str.getvalue()
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:33]: Why is this necessary? The idiomatic Python way to build a string is to use a list to store intermediate strings and join at the end.

File: scripts/gen_code_completion.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+def gen_keyword_list(file) -> list:
+    p = re.compile(r'(?<=keyword yagpdbcc)\w+\s+(.*?)(?=\s+contained$)', re.MULTILINE)
+    with open(file) as f:
+        matches = [match for line in f if (match := p.findall(line))]
+        keywords = []
+        for sublist in matches:
+            for item in sublist:
+                tmp = item.split()
+                for keyw in tmp:
+                    keywords.append(keyw)
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:44]:
```suggestion
                yield from item.split()
```

File: scripts/gen_code_completion.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+def gen_completion(keywords) -> str:
+    string_builder = StringBuilder()
+
+    for keyword in keywords:
+        string_builder.Add(f"\t\t\u007b label = '{keyword}' \u007d,\n")
+
+    return string_builder.__str__()
+
+def write_file(code):
+    if not os.listdir().__contains__('lua'):
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:57]: Is there a reason this isn't simply
```suggestion
    if "lua" not in os.listdir():
```

File: scripts/gen_code_completion.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        try:
+            os.mkdir('lua')
+        except OSError as error:
+            print(error)
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:61]: This should be a fatal error, no?

File: scripts/gen_code_completion.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    with open('scripts/boilerplate.lua', "rt") as boilerplate:
+        with open('lua/yagpdbcc.lua', "wt") as fout:
+            for line in boilerplate:
+                fout.write(line.replace('---@@cmp-src@@', code))
+
+def main():
+    print("Generating sources...")
+
+    path = r'syntax/yagpdbcc/'
+    keywords = list()
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:72]:
```suggestion
    keywords = []
```

File: scripts/gen_code_completion.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self):
+        self.file_str = StringIO()
+
+    def Add(self, str):
+        self.file_str.write(str)
+
+    def __str__(self):
+        return self.file_str.getvalue()
+
+def gen_keyword_list(file) -> list:
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:35]: Instead of using a list, you can use a generator instead, which *generates* items lazily on demand (hence the name.)

File: scripts/gen_code_completion.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    with open(file) as f:
+        matches = [match for line in f if (match := p.findall(line))]
+        keywords = []
+        for sublist in matches:
+            for item in sublist:
+                tmp = item.split()
+                for keyw in tmp:
+                    keywords.append(keyw)
+
+    return keywords
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:46]:
```suggestion
```

File: scripts/gen_code_completion.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+def main():
+    print("Generating sources...")
+
+    path = r'syntax/yagpdbcc/'
+    keywords = list()
+
+    with os.scandir(path) as dirs:
+        for entry in dirs:
+            if entry.name.endswith('.vim'):
+                keywords.append(gen_keyword_list(path + entry.name))
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:77]:
```suggestion
                keywords.extend(gen_keyword_list(path + entry.name))
```

File: scripts/gen_code_completion.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+    # Flatten the keyword list again because calling the above twice causes a
+    # nested list, which is undesirable.
+    keywords = [keyword for keywords in keywords for keyword in keywords]
```
Comment (jo3-l) [inline scripts/gen_code_completion.py:81]:
```suggestion
```

**Comment by jo3-l** [comment]
You may want to view the suggestions in a linear fashion instead of in the order GitHub shows them.

**Comment by l-zeuch** [comment]
Thanks for the feedback, I addressed it in 4262c92!

File: scripts/gen_code_completion.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+def write_file(code):
+    if "lua" not in os.listdir():
+        try:
+            os.mkdir('lua')
+        except OSError as error:
+            print(error)
+            os._exit(1)
+
+    with open('scripts/boilerplate.lua', "rt") as boilerplate:
+        with open('lua/yagpdbcc.lua', "wt") as fout:
```
Comment (LRitzdorf) [inline scripts/gen_code_completion.py:44]: I like the Pythonicity here, but we can condense this to `with open(...) as biolerplate, open(...) as fout:`. Just to flatten the structure a bit

File: lua/yagpdbcc.lua [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+function source:get_debug_name()
+    return 'yagpdb-cc'
+end
+
+---Invoke completion. (Required)
+---@param params cmp.SourceCompletionApiParams
+---@param callback fun(response: lsp.CompletionResponse|nil)
+function source:complete(params, callback)
+    callback({
+		{ label = 'true' },
```
Comment (LRitzdorf) [inline lua/yagpdbcc.lua:44]: On the one hand, committing the generated completions makes sense — users will be able to jump right in. On the other hand, we have a `make` target for this, which *usually* means you don't commit the products...

Personally, I do think it makes more sense to commit the completions, but we have to be sure to update them along with the rest of the code. Maybe their generation should be automated as part of some kind of release pipeline?

**Review by LRitzdorf — COMMENTED** [review]
Looking mostly good! A few quick comments before we merge, though

File: lua/yagpdbcc.lua (hunk 11 above)
Comment (l-zeuch) [reply lua/yagpdbcc.lua:44]: Sure, I was thinking about either crontabbing or running generation when any file in `syntax/` changed, if not both. Though crontab seems rather overkill...

As a user, I would expect an out-of-the-box experience with regards to promoted features, so I'd say we commit the result.

If the user wants to patch something in, then they can use the `make` target for convenience.

**Review by LRitzdorf — APPROVED** [review]
The automatic completion generation looks great!

**Comment by l-zeuch** [comment]
Let's :shipit:
